### PR TITLE
demisto/etc#23233 added all files validation

### PR DESCRIPTION
**Pull request by hod-alpert**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: demisto/etc#23233

## Description
Adds an option to add full validation to all files using `demisto-sdk validate -a`

## Must have
- [ ] Tests
- [x] Documentation

### CODE REVIEW

File: demisto_sdk/__main__.py [hunk 1]
```diff
@@ -204,6 +204,9 @@ def unify(**kwargs):
 @click.option(
     '-p', '--path', help='Path of file to validate specifically, outside of a git directory.'
 )
+@click.option(
+    '-a', '--validate-all', is_flag=True, show_default=True, default=False, help='Whether to validate all files or not'
```
Comment (Itay4) [inline demisto_sdk/__main__.py:208]: lets improve description here, and add to it that it will run all validations

File: demisto_sdk/commands/validate/file_validator.py [hunk 2]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
 
     def validate_pack_unique_files(self, packs):
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:528]: please add type hints and docstring here

File: demisto_sdk/commands/validate/file_validator.py [hunk 3]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
         for pack in packs:
+            print(f'Validating {pack} unique pack files')
             pack_unique_files_validator = PackUniqueFilesValidator(pack)
             pack_errors = pack_unique_files_validator.validate_pack_unique_files()
             if pack_errors:
                 print_error(pack_errors)
                 self._is_valid = False
 
+    def file_all_validations(self, file_path, file_type: str = None):
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:537]: please add type hints and docstring here

File: demisto_sdk/commands/validate/file_validator.py [hunk 4]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+                        'Incident fields, Indicator fields, Images, Release notes, Layouts and Descriptions')
+            self._is_valid = False
+
     def validate_all_files(self):
+        print('Validating all files')
+        print('Validating conf.json')
+        conf_json_validator = ConfJsonValidator()
+        if not conf_json_validator.is_valid_conf_json():
+            self._is_valid = False
+        packs = {os.path.basename(pack) for pack in glob('Packs/*')}
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:635]:
```suggestion
        packs = {os.path.basename(pack) for pack in glob(f'{PACKS_DIR}/*')}
```

File: demisto_sdk/commands/validate/file_validator.py [hunk 5]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+        self.validate_pack_unique_files(packs)
+        all_files = set()
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:637]:
```suggestion
        all_files_to_validate = set()
```

File: demisto_sdk/commands/validate/file_validator.py [hunk 6]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+        for directory in [PACKS_DIR, BETA_INTEGRATIONS_DIR]:
+            all_files |= {file for file in glob(f'{directory}/**', recursive=True) if not os.path.isdir(file)}
+        print('Validating pack files')
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:640]: improve message here to be more indicative (beta integration related)

File: demisto_sdk/commands/validate/file_validator.py (hunk 3 above)
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:537]:
```suggestion
    def run_all_validations_on_file(self, file_path, file_type: str = None):
```

File: demisto_sdk/commands/validate/file_validator.py [hunk 7]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
         for pack in packs:
+            print(f'Validating {pack} unique pack files')
             pack_unique_files_validator = PackUniqueFilesValidator(pack)
             pack_errors = pack_unique_files_validator.validate_pack_unique_files()
             if pack_errors:
                 print_error(pack_errors)
                 self._is_valid = False
 
+    def file_all_validations(self, file_path, file_type: str = None):
+        file_extention = os.path.splitext(file_path)[-1]
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:538]:
```suggestion
        file_extension = os.path.splitext(file_path)[-1]
```

File: demisto_sdk/commands/validate/file_validator.py [hunk 8]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+        if re.match(TEST_PLAYBOOK_REGEX, file_path, re.IGNORECASE) and not file_type:
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:539]: for every return, add comment why

File: demisto_sdk/commands/validate/file_validator.py [hunk 9]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+            return
+        if file_extention not in ['.yml', 'json', '.md']:
+            return
+        # Ignoring changelog and description files
+        if 'changelog' in file_path.lower() or 'description' in file_path.lower():
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:544]: lets do `file_path.lower()` once

File: demisto_sdk/commands/validate/file_validator.py (hunk 8 above)
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:539]: check if the `and` part is needed

File: demisto_sdk/commands/validate/file_validator.py [hunk 10]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+        print(f'Validating {file_path}')
+        if re.match(TEST_PLAYBOOK_REGEX, file_path, re.IGNORECASE):
+            return
+
+        elif 'README' in file_path:
+            readme_validator = ReadMeValidator(file_path)
+            if not readme_validator.is_valid_file():
+                self._is_valid = False
+            return
+        elif 'description' in file_path.lower():
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:555]: check if needed

File: demisto_sdk/commands/validate/file_validator.py [hunk 11]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+            if not script_validator.is_valid_file(validate_rn=False):
+                self._is_valid = False
+
+        elif re.match(BETA_INTEGRATION_REGEX, file_path, re.IGNORECASE) or \
+                re.match(BETA_INTEGRATION_YML_REGEX, file_path, re.IGNORECASE) or file_type == 'betaintegration':
+            integration_validator = IntegrationValidator(structure_validator)
+            if not integration_validator.is_valid_beta_integration():
+                self._is_valid = False
+
+        elif re.match(IMAGE_REGEX, file_path, re.IGNORECASE):
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:588]: check if you can remove this

File: demisto_sdk/commands/validate/file_validator.py [hunk 12]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+            if not dashboard_validator.is_valid_dashboard(validate_rn=False):
+                self._is_valid = False
+
+        elif checked_type(file_path, JSON_ALL_INCIDENT_TYPES_REGEXES) or file_type == 'incidenttype':
+            incident_type_validator = IncidentTypeValidator(structure_validator)
+            if not incident_type_validator.is_valid_incident_type(validate_rn=False):
+                self._is_valid = False
+
+        elif checked_type(file_path, CHECKED_TYPES_REGEXES):
+            pass
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:621]: add print here like "found type ... but has no validations for it"

File: demisto_sdk/commands/validate/file_validator.py [hunk 13]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+        elif 'description' in file_path.lower():
+            description_validator = DescriptionValidator(file_path)
+            if not description_validator.is_valid():
+                self._is_valid = False
+                return
+        structure_validator = StructureValidator(file_path, predefined_scheme=file_type)
+        if not structure_validator.is_valid_file():
+            self._is_valid = False
+
+        elif re.match(PLAYBOOK_REGEX, file_path, re.IGNORECASE) or file_type == 'playbook':
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:564]: what about pack/beta playbooks?

File: demisto_sdk/commands/validate/file_validator.py (hunk 8 above)
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:539]: duplicate as 547

File: demisto_sdk/commands/validate/file_validator.py [hunk 14]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+        file_extention = os.path.splitext(file_path)[-1]
+        if re.match(TEST_PLAYBOOK_REGEX, file_path, re.IGNORECASE) and not file_type:
+            return
+        if file_extention not in ['.yml', 'json', '.md']:
+            return
+        # Ignoring changelog and description files
+        if 'changelog' in file_path.lower() or 'description' in file_path.lower():
+            return
+        print(f'Validating {file_path}')
+        if re.match(TEST_PLAYBOOK_REGEX, file_path, re.IGNORECASE):
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:547]: should include minimal validator checks (`id == name`, `version == -1`)

Comment (yaakovi) [reply demisto_sdk/commands/validate/file_validator.py:547]: this can be improved in another PR

File: demisto_sdk/commands/validate/file_validator.py [hunk 15]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+        structure_validator = StructureValidator(file_path, predefined_scheme=file_type)
+        if not structure_validator.is_valid_file():
+            self._is_valid = False
+
+        elif re.match(PLAYBOOK_REGEX, file_path, re.IGNORECASE) or file_type == 'playbook':
+            playbook_validator = PlaybookValidator(structure_validator)
+            if not playbook_validator.is_valid_playbook():
+                self._is_valid = False
+
+        elif checked_type(file_path, YML_INTEGRATION_REGEXES) or file_type == 'integration':
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:569]: `YML_INTEGRATION_REGEXES` also include beta regex so this might conflict with line 582

File: demisto_sdk/commands/validate/file_validator.py [hunk 16]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+
+        elif checked_type(file_path, PACKAGE_SCRIPTS_REGEXES) or file_type == 'script':
+            # Set file path to the yml file
+            structure_validator.file_path = file_path
+            script_validator = ScriptValidator(structure_validator)
+
+            if not script_validator.is_valid_file(validate_rn=False):
+                self._is_valid = False
+
+        elif re.match(BETA_INTEGRATION_REGEX, file_path, re.IGNORECASE) or \
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:582]: use checked_type with `YML_BETA_INTEGRATIONS_REGEXES`

File: demisto_sdk/commands/validate/file_validator.py [hunk 17]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+                re.match(BETA_INTEGRATION_YML_REGEX, file_path, re.IGNORECASE) or file_type == 'betaintegration':
+            integration_validator = IntegrationValidator(structure_validator)
+            if not integration_validator.is_valid_beta_integration():
+                self._is_valid = False
+
+        elif re.match(IMAGE_REGEX, file_path, re.IGNORECASE):
+            image_validator = ImageValidator(file_path)
+            if not image_validator.is_valid():
+                self._is_valid = False
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:591]: same as description
unreachable for png due to line 541

File: demisto_sdk/commands/validate/file_validator.py (hunk 12 above)
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:621]: when will you get here?

File: demisto_sdk/commands/validate/file_validator.py [hunk 18]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+            if not incident_type_validator.is_valid_incident_type(validate_rn=False):
+                self._is_valid = False
+
+        elif checked_type(file_path, CHECKED_TYPES_REGEXES):
+            pass
+
+        else:
+            print_error('The file type of {} is not supported in validate command'.format(file_path))
+            print_error('validate command supports: Integrations, Scripts, Playbooks, '
+                        'Incident fields, Indicator fields, Images, Release notes, Layouts and Descriptions')
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:626]: dashboards, incident types, reputations

File: demisto_sdk/commands/validate/file_validator.py [hunk 19]
```diff
@@ -523,13 +527,121 @@ def validate_committed_files(self):
...
+            playbook_validator = PlaybookValidator(structure_validator)
+            if not playbook_validator.is_valid_playbook():
+                self._is_valid = False
+
+        elif checked_type(file_path, YML_INTEGRATION_REGEXES) or file_type == 'integration':
+            integration_validator = IntegrationValidator(structure_validator)
+            if not integration_validator.is_valid_file(validate_rn=False):
+                self._is_valid = False
+
+        elif checked_type(file_path, PACKAGE_SCRIPTS_REGEXES) or file_type == 'script':
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:574]:
```suggestion
        elif checked_type(file_path, YML_ALL_SCRIPTS_REGEXES) or file_type == 'script':
```

this does not include a unified script in packs, please add

File: demisto_sdk/commands/validate/file_validator.py (hunk 9 above)
Comment (hod-alpert) [reply demisto_sdk/commands/validate/file_validator.py:544]: We cannot because some validators should open the file

File: demisto_sdk/commands/validate/file_validator.py (hunk 13 above)
Comment (hod-alpert) [reply demisto_sdk/commands/validate/file_validator.py:564]: beta playbooks will be removed to legacy playbooks

Packs playbooks are validated

File: demisto_sdk/commands/validate/file_validator.py (hunk 14 above)
Comment (hod-alpert) [reply demisto_sdk/commands/validate/file_validator.py:547]: I'll have a commit for that

File: demisto_sdk/commands/validate/file_validator.py (hunk 13 above)
Comment (yaakovi) [reply demisto_sdk/commands/validate/file_validator.py:564]: where are they validated?

File: demisto_sdk/commands/validate/file_validator.py (hunk 12 above)
Comment (hod-alpert) [reply demisto_sdk/commands/validate/file_validator.py:621]: Not sure, copied from validate_added_files, added a print here, so I'll see what happens

File: demisto_sdk/commands/validate/file_validator.py (hunk 13 above)
Comment (hod-alpert) [reply demisto_sdk/commands/validate/file_validator.py:564]: [this](https://github.com/demisto/demisto-sdk/blob/4093d6b60fab342418ae0c439de0a2bad9655c94/demisto_sdk/commands/validate/file_validator.py#L639) catches all files in Packs recursively

Comment (yaakovi) [reply demisto_sdk/commands/validate/file_validator.py:577]: but then you get here and ignore them since they are not accepted by this regex

File: demisto_sdk/commands/common/hook_validations/image.py [hunk 20]
```diff
@@ -119,8 +117,12 @@ def load_image_from_yml(self):
...
 
     def load_image(self):
-        if re.match(IMAGE_REGEX, self.file_path, re.IGNORECASE):
-            with open(self.file_path, "rb") as image:
+        package_path = os.path.dirname(self.file_path)
+        image_paths = glob.glob(package_path + '/*.png')
+        if image_paths:
+            image_path = next(image for image in image_paths)
+            with open(image_path, "rb") as image:
+                # TODO talk to blob about this
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/image.py:125]: ?

File: demisto_sdk/commands/validate/file_validator.py [hunk 21]
```diff
@@ -525,7 +524,16 @@ def validate_committed_files(self):
             self.validate_no_old_format(old_format_files)
             self.validate_pack_unique_files(packs)
 
-    def validate_pack_unique_files(self, packs):
+    def validate_pack_unique_files(self, packs: set) -> None:
+        """
+        Runs validations on the following pack files:
+        * .secret-ignore: Validates that the file exist and that the file's secrets can be parsed as a list delimited by '\n'
+        * .pack-ignore: Validates that the file exists and that all regexes in it can be compiled
+        * README..md file: Validates that the file exists
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:532]:
```suggestion
        * README.md file: Validates that the file exists
```

File: demisto_sdk/commands/validate/file_validator.py [hunk 22]
```diff
@@ -525,7 +524,16 @@ def validate_committed_files(self):
...
+        * pack_metadata: Validates that the file exists and that it has a valid structure
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py:533]:
```suggestion
        * pack_metadata.json: Validates that the file exists and that it has a valid structure
```

File: demisto_sdk/commands/common/constants.py [hunk 23]
```diff
@@ -444,6 +445,7 @@ def found_hidden_param(parameter_name):
 INTEGRATION_PS_REGEX = r'{}{}/([^\\/]+)/\1.ps1$'.format(CAN_START_WITH_DOT_SLASH, INTEGRATIONS_DIR)
 INTEGRATION_YML_REGEX = r'{}{}/([^\\/]+)/([^\\/]+).yml$'.format(CAN_START_WITH_DOT_SLASH, INTEGRATIONS_DIR)
 INTEGRATION_REGEX = r'{}{}/(integration-[^\\/]+)\.yml$'.format(CAN_START_WITH_DOT_SLASH, INTEGRATIONS_DIR)
+LEGACY_INTEGRATION_REGEX = r'{}{}/(integration-[^\\/]+)\.yml$'.format(CAN_START_WITH_DOT_SLASH, LEGACY_INTEGRATIONS_DIR)
```
Comment (yaakovi) [inline demisto_sdk/commands/common/constants.py:448]: where are you using this?

File: demisto_sdk/commands/validate/file_validator.py (hunk 13 above)
Comment (hod-alpert) [reply demisto_sdk/commands/validate/file_validator.py:577]: The regexes are broken and need to have a robust fix.
The playbooks are not ignored because of the `or file_type == 'playbook'` part

File: demisto_sdk/commands/common/schemas/playbook.yml [hunk 24]
```diff
@@ -1,5 +1,7 @@
 type: map
 mapping:
+  beta:
```
Comment (Itay4) [inline demisto_sdk/commands/common/schemas/playbook.yml:3]: why did you add that? there is no beta playbook

File: demisto_sdk/commands/common/tools.py [hunk 25]
```diff
@@ -605,8 +605,8 @@ def find_type(path: str):
                 return 'layout'
             else:
                 return 'dashboard'
-
-        elif 'id' in _dict:
+        # When using it for all files validation- sometimes 'id' can be integer
+        elif 'id' in _dict and isinstance(_dict.get('id'), str):
```
Comment (Itay4) [inline demisto_sdk/commands/common/tools.py:609]: what will happen in case it is an integer? what will the validation print?

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 26]
```diff
@@ -8,18 +8,26 @@
...
+            is_test_playbook (bool): whether the playbook is a test playbook
             is_new_playbook (bool): whether the playbook is new or modified
             validate_rn (bool):  whether we need to validate release notes or not
 
         Returns:
             bool. Whether the playbook is valid or not
         """
-
-        if is_new_playbook:
+        if is_test_playbook:
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/playbook.py:24]: why do we need special case for test playbooks?

File: demisto_sdk/commands/validate/tests/validators_test.py [hunk 27]
```diff
@@ -86,6 +87,41 @@ def teardown_class(cls):
...
 
+    INPUTS_IS_VALID_TEST_PLAYBOOKS = [
+        (VALID_TEST_PLAYBOOK_PATH, True),
+        (INVALID_PLAYBOOK_PATH, False),
+        (INVALID_PLAYBOOK_PATH_FROM_ROOT, True)
+    ]
+
+    @pytest.mark.parametrize('source, answer', INPUTS_IS_VALID_TEST_PLAYBOOKS)
+    def test_validation_of_test_playbooks(self, source, answer):
+        """
```
Comment (Itay4) [inline demisto_sdk/commands/validate/tests/validators_test.py:98]: for all tests you added, update the test docstring according to `Given - When - Then` template

File: demisto_sdk/commands/validate/tests/validators_test.py [hunk 28]
```diff
@@ -335,6 +371,34 @@ def test_is_valid_rn(self, mocker, file_path, file_type):
...
+    FILES_PATHS_FOR_ALL_VALIDATIONS = [
+        # ignoring images and change-logs
+        (DEFAULT_IMAGE, ''),
+        (VALID_MULTI_LINE_LIST_CHANGELOG_PATH, ''),
+        (INVALID_ONE_LINE_1_CHANGELOG_PATH, ''),
+        # validating files
+        (VALID_INTEGRATION_ID_PATH, 'integration'),
+        (VALID_TEST_PLAYBOOK_PATH, 'playbook'),
+        (VALID_SCRIPT_PATH, 'script'),
+        (VALID_DASHBOARD_PATH, 'dashboard'),
```
Comment (Itay4) [inline demisto_sdk/commands/validate/tests/validators_test.py:383]: what about beta integration?

File: demisto_sdk/commands/common/schemas/playbook.yml (hunk 24 above)
Comment (hod-alpert) [reply demisto_sdk/commands/common/schemas/playbook.yml:3]: There's `Beta_Integrations/playbook-TrendMicro_DDA_DetonateURL.yml` for example

there are a few more so I figured I'd add it

File: demisto_sdk/commands/common/tools.py (hunk 25 above)
Comment (hod-alpert) [reply demisto_sdk/commands/common/tools.py:609]: I till raise error because in the next line we do `id.lower()`

File: demisto_sdk/commands/common/schemas/playbook.yml (hunk 24 above)
Comment (Itay4) [reply demisto_sdk/commands/common/schemas/playbook.yml:3]: not sure why that playbook is there, but there is no such thing as beta playbook, so lets remove that

File: demisto_sdk/commands/common/tools.py (hunk 25 above)
Comment (Itay4) [reply demisto_sdk/commands/common/tools.py:609]: im ok with an error, just asking what it looks like? is it indicative?

File: demisto_sdk/commands/common/hook_validations/playbook.py (hunk 26 above)
Comment (hod-alpert) [reply demisto_sdk/commands/common/hook_validations/playbook.py:24]: Not sure about the specifics but all validators simply skip test playbooks for some reason, If I recall correctly @yaakovi told me this is a check we can add instead of skipping them.

Comment (hod-alpert) [reply demisto_sdk/commands/common/hook_validations/playbook.py:24]: image

Comment (Itay4) [reply demisto_sdk/commands/common/hook_validations/playbook.py:24]: spoke with @yaakovi, sorry for the confusion but no real need to have special case for test playbooks, id prefer we stay consistent as there is no real reason not to run the same validations

File: demisto_sdk/commands/validate/tests/validators_test.py (hunk 28 above)
Comment (hod-alpert) [reply demisto_sdk/commands/validate/tests/validators_test.py:383]: Added, although there is no beta-integration schema and the function `find_type` returns `integration` for beta integrations.

File: demisto_sdk/commands/common/schemas/playbook.yml (hunk 24 above)
Comment (hod-alpert) [reply demisto_sdk/commands/common/schemas/playbook.yml:3]: But then it fails and breaks the whole validation process

Comment (hod-alpert) [reply demisto_sdk/commands/common/schemas/playbook.yml:3]: There are four of them and they will fail the build

File: demisto_sdk/commands/common/tools.py (hunk 25 above)
Comment (hod-alpert) [reply demisto_sdk/commands/common/tools.py:609]: image

Comment (hod-alpert) [reply demisto_sdk/commands/common/tools.py:609]: It's not indicative and it's unexpected so it breaks the validation process
image

Comment (Itay4) [reply demisto_sdk/commands/common/tools.py:609]: ok lets handle that in the `else` clause of that `elif`
